kymatio/scattering2d/backend/torch_backend.py

```diff
@@ -0,0 +1,346 @@
+# Authors: Edouard Oyallon, Sergey Zagoruyko
+
+import torch
+from torch.nn import ReflectionPad2d
+from collections import namedtuple
+
+BACKEND_NAME = 'torch'
+
+def iscomplex(x):
+    return x.shape[-1] == 2
+
+def isreal(x):
+    return x.shape[-1] == 1
+
+class Pad(object):
+    def __init__(self, pad_size, input_size, pre_pad=False):
+        """Padding which allows to simultaneously pad in a reflection fashion
+            and map to complex.
+
+            Parameters
+            ----------
+            pad_size : list of 4 integers
+                Size of padding to apply [top, bottom, left, right].
+            input_size : list of 2 integers
+                size of the original signal [height, width].
+            pre_pad : boolean, optional
+                If set to true, then there is no padding, one simply adds the imaginary part.
+
+            Attributes
+            ----------
+            pad_size : list of 4 integers 
+                Size of padding to apply [top, bottom, left, right].
+            input_size : list of 2 integers
+                Size of the original signal [height, width].
+            pre_pad : boolean
+                If set to true, then there is no padding, one simply adds the imaginary part.
+                
+        """
+        self.pre_pad = pre_pad
+        self.pad_size = pad_size
+        self.input_size = input_size
+
+        self.build()
+
+    def build(self):
+        """Builds the padding module.
+
+            Attributes
+            ----------
+            padding_module : ReflectionPad2d
+                Pads the input tensor using the reflection of the input
+                boundary. 
+
+        """
+        pad_size_tmp = list(self.pad_size)
+
+        # This handles the case where the padding is equal to the image size
+        if pad_size_tmp[0] == self.input_size[0]:
+            pad_size_tmp[0] -= 1
+            pad_size_tmp[1] -= 1
+        if pad_size_tmp[2] == self.input_size[1]:
+            pad_size_tmp[2] -= 1
+            pad_size_tmp[3] -= 1
+        # Pytorch expects its padding as [left, right, top, bottom]
+        self.padding_module = ReflectionPad2d([pad_size_tmp[2], pad_size_tmp[3],
+                                               pad_size_tmp[0], pad_size_tmp[1]])
+
+    def __call__(self, x):
+        """Applies padding and maps to complex.
+            
+            Parameters
+            ----------
+            x : tensor
+                Real tensor input to be padded and sent to complex domain.
+
+            Returns
+            -------
+            output : tensor
+                Complex torch tensor that has been padded.
+
+        """
+        batch_shape = x.shape[:-2]
+        signal_shape = x.shape[-2:]
+        x = x.reshape((-1, 1) + signal_shape)
+        if not self.pre_pad:
+            x = self.padding_module(x)
+            if self.pad_size[0] == self.input_size[0]:
```

Comment:
@edouardoyallon

Reply:
@Kegnarok ?

AFAIR, it's a way to handle the case N=2^J which was not handled well.

Reply:
This is manual reflection padding because Pytorch can't handle it for arbitrary sizes:
> The problem is that reflection padding isn't defined for a padding equal to the input size: you can extend `0 1 2` to `2 1 | 0 1 2 | 1 0` but no further. So in order to add that extra padding we manually do a second reflection: `1 2 1 | 0 1 2 | 1 0 1`.
(see https://github.com/kymatio/kymatio/pull/412/files/9885f411f914de080ef69fe458ccfd42b33025f4#r305372164)

Reply:
Ok. Can we put in a comment to that effect?

Reply:
Agreed, let's put a comment `# Note: pytorch is not effective to pad signals of size N-1 with N elements, thus we had to add this fix`.

Reply:
Done.